### ISSUE TITLE
CuDNN RNN bindings are basically all deprecated in cudnn 8

### ISSUE THREAD

**Issue by mcarilli** [pytorch/pytorch#42545]
Many (all?!!) the raw cudnn calls in
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cudnn/Descriptors.h
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cudnn/RNN.cpp
are deprecated according to
https://docs.nvidia.com/deeplearning/cudnn/api/index.html#api-changes

Not house on fire, we have until cudnn 9 to update, filing so we don't lose track.  I don't mind taking this on myself since my mental cache is full of RNN stuff after https://github.com/pytorch/pytorch/pull/42385, but I can't work on it immediately.

**Comment by ezyang** [comment]
Yes, this would be great to update. Although, I don't know how different the new API calls are; maybe we will have to rewrite all of RNN.cpp

**Comment by mcarilli** [comment]
Tbh I haven't fully scoped this.  Some are drop-in replacements, others fuse certain existing calls.  For example, [cudnnGetRNNLinLayerMatrixParams](https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnGetRNNLinLayerMatrixParams) and [cudnnGetRNNLinLayerBiasParams](https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnGetRNNLinLayerBiasParams) from 7.x are unified into [cudnnGetRNNWeightParams](https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnGetRNNWeightParams) in 8.x, so [get_parameters](https://github.com/pytorch/pytorch/blob/b08347fd7b305e616de1ddb964747ce3fb57c788/aten/src/ATen/native/cudnn/RNN.cpp#L411) needs a rewrite.  The rewrite could be shorter, which is nice, but the code would be divergent, so while supporting both 7.x and 8.x we would need two different hardcoded implementations selected by cudnn version macros.
